util/setuidgid.c

```diff
@@ -0,0 +1,360 @@
+
+#include <assert.h>
+#include <errno.h>
+#include <grp.h>
+#include <pwd.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "warnp.h"
+
+#include "setuidgid.h"
+
+/* Function prototypes related to supplementary groups. */
+static int setgroups_none(void);
+static int check_supplementary_groups_none(void);
+
+/* Function prototypes related to uid and gid. */
+static int set_group(const char *);
+static int set_user(const char *);
+static int string_extract_user_group(const char *, char **, char **);
+
+/**
+ * setgroups_none():
+ * Leave all supplementary groups.  If we do not know how to do this on the
+ * platform, return 0 anyway.
+ */
+static int
+setgroups_none()
+{
+
+#ifdef HAS_SETGROUPS
+	if (setgroups(0, NULL))
+		goto err0;
+#endif
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Check if we're in any supplementary groups. */
+static int
+check_supplementary_groups_none()
+{
+	gid_t grouplist[1];
+	int ngroups;
+
+	/* Find number of groups. */
+	if ((ngroups = getgroups(0, NULL)) < 0) {
+		warnp("getgroups()");
+		goto err0;
+	}
+
+	/* Check group membership. */
+	if (ngroups > 1)
+		/* Definitely failed. */
+		goto err0;
+	if (ngroups == 1) {
+		/*
+		 * POSIX allows getgroups() to return the effective group ID,
+		 * so if we're in exactly one group, we need to check that GID.
+		 */
+		if ((ngroups = getgroups(1, grouplist)) < 0) {
+			warnp("getgroups()");
+			goto err0;
+		}
+		if (grouplist[0] != getegid())
+			goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
+ * set_group(groupname):
+ * Set the effective GID to the group represented by ${groupname}.
+ */
+static int
+set_group(const char * groupname)
+{
+	gid_t gid;
+	struct group * group_info;
+
+	/* Find numerical GID. */
+	errno = 0;
+	if ((group_info = getgrnam(groupname)) == NULL) {
+		warnp("getgrnam(%s)", groupname);
+		goto err0;
+	}
+	gid = group_info->gr_gid;
+
+	/* Set GID. */
+	if (setgid(gid) != 0) {
+		if (errno == EPERM)
+			goto nopermission;
+		warnp("Failed to set GID");
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+nopermission:
+	errno = EPERM;
+
+	/* Failure! */
+	return (-1);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
+ * set_user(username):
+ * Set the effective UID to the user represented by ${username}.
+ */
+static int
+set_user(const char * username)
+{
+	uid_t uid;
+	struct passwd * user_info;
+
+	/* Calculate numerical UID. */
+	errno = 0;
+	if ((user_info = getpwnam(username)) == NULL) {
+		warn0("getpwnam(%s)", username);
+		goto err0;
+	}
+	uid = user_info->pw_uid;
+
+	/* Set UID. */
+	if (setuid(uid) != 0) {
+		if (errno == EPERM)
+			goto nopermission;
+		warnp("Failed to set UID");
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+nopermission:
+	errno = EPERM;
+
+	/* Failure! */
+	return (-1);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Parses "username" or "username:groupgroup". */
```

Comment:
We should also accept ":groupname", meaning "change the group but not the user".

Also, s/groupgroup/groupname/ ?